spinalcordtoolbox/csa_pmj.py

```diff
@@ -0,0 +1,179 @@
+#!/usr/bin/env python
+# -*- coding: utf-8
+# Functions to get distance from PMJ for processing segmentation data
+# Author: Sandrine Bédard
+import logging
+import numpy as np
+
+from spinalcordtoolbox.image import Image
+from spinalcordtoolbox.centerline.core import get_centerline
+from spinalcordtoolbox.resampling import resample_nib
+
+logger = logging.getLogger(__name__)
+
+NEAR_ZERO_THRESHOLD = 1e-6
+
+
+def get_slices_for_pmj_distance(segmentation, pmj, distance, extent, param_centerline=None, verbose=1):
+    """
+    Compute distance from PMJ projection on centerline for all the centerline.
+    Generate mask from segmentation of the slices used to process segmentation data corresponding to a distance from PMJ projection.
+    :param segmentation: input segmentation. Could be either an Image or a file name.
+    :param pmj: label of PMJ.
+    :param distance: float: Distance from PMJ.
+    :param param_centerline: see centerline.core.ParamCenterline()
+    :param verbose:
+    :return mask:
+    :return slices:
+
+    """
+    # TODO: loop through distance (when multiple distances are given)
+    im_seg = Image(segmentation).change_orientation('RPI')
+    im_pmj = Image(pmj).change_orientation('RPI')
+    nx, ny, nz, nt, px, py, pz, pt = im_seg.dim
+    pr = min([px, py])
+    im_segr = resample_nib(im_seg, new_size=[pr, pr, pz], new_size_type='mm', interpolation='linear')
+    im_pmjr = resample_nib(im_pmj, new_size=[pr, pr, pz], new_size_type='mm', interpolation='linear')
+    data_pmj = im_pmjr.data
+    # Update dimensions from resampled image
+    nx, ny, nz, nt, px, py, pz, pt = im_segr.dim
+
+    # Extract min and max index in Z direction
+    data_seg = im_segr.data
+    X, Y, Z = (data_seg > NEAR_ZERO_THRESHOLD).nonzero()
+    min_z_index, max_z_index = min(Z), max(Z)
+
+    # Compute the spinal cord centerline based on the spinal cord segmentation
+    _, arr_ctl, arr_ctl_der, fit_results = get_centerline(im_segr, param=param_centerline, verbose=verbose)
+
+    # Get tagent vector at the top of the centerline | Maybe take a fiew slices to get the vector ?
+    iz = max_z_index
+    tangent_vect = np.array([arr_ctl_der[0][iz - min_z_index] * px, arr_ctl_der[1][iz - min_z_index] * py, pz])
```

Comment:
You got the intuition right that "take a few slices to get the vector" would be more robust to issues in the latest segmentation, but even so, I think that taking the tangent is not a robust approach as it is way to sensitive to the top segmented slices, which are usually only partially segmented. To demonstrate the sensitivity, I slightly modified the top two slices of the segmentation produced by `sct_deepseg_sc` on the T2w scan from the **sct_example_data**. The original segmentation is in white, and the modified one in red:
![image](https://user-images.githubusercontent.com/2482071/122834223-2caaae80-d2bc-11eb-8689-2706cd82461d.png)

When running the following command on the original segmentation:
~~~
sct_process_segmentation -i t2_seg.nii.gz -pmj t2_pmj.nii.gz -distance 55 -extent 20 -o csa_pmj.csv
~~~

I get this:


Timestamp | SCT Version | Filename | Slice (I->S) | VertLevel | DistancePMJ | MEAN(area) | STD(area) | MEAN(angle_AP) | STD(angle_AP) | MEAN(angle_RL) | STD(angle_RL) | MEAN(diameter_AP) | STD(diameter_AP) | MEAN(diameter_RL) | STD(diameter_RL) | MEAN(eccentricity) | STD(eccentricity) | MEAN(orientation) | STD(orientation) | MEAN(solidity) | STD(solidity) | SUM(length)
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
2021-06-21 13:52:13 | git-sb/3063-csa-measure-from-pmj-122c615148a38e8550a0ae9e1d4ed1375857015b | /Users/julien/sct_example_data/t2/t2_seg.nii.gz | 176:200 |   | 55.0 | 73.09505206059430 | 2.058273644130360 | 0.6042205187778950 | 0.509096773612399 | 5.99722237889052 | 1.3648000670811700 | 8.044099345457310 | 0.2133923010218110 | 11.551670079850800 | 0.11679222483932000 | 0.7170797849523140 | 0.019380548885043800 | 1.1437363154081200 | 1.0298888376877800 | 0.9723314116348530 | 0.009082028290704460 | 20.117826899911600

While when running it on the modified segmentation, I get a very different selection of slices (shifted by 10 slices, which represents 8.0 mm), even though the distance to the PMJ is the same:



Timestamp | SCT Version | Filename | Slice (I->S) | VertLevel | DistancePMJ | MEAN(area) | STD(area) | MEAN(angle_AP) | STD(angle_AP) | MEAN(angle_RL) | STD(angle_RL) | MEAN(diameter_AP) | STD(diameter_AP) | MEAN(diameter_RL) | STD(diameter_RL) | MEAN(eccentricity) | STD(eccentricity) | MEAN(orientation) | STD(orientation) | MEAN(solidity) | STD(solidity) | SUM(length)
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
2021-06-21 18:06:28 | git-sb/3063-csa-measure-from-pmj-0696bb0abe42ee8b9a9bf72ecb87a467092ccd7e | /Users/julien/sct_example_data/t2/t2_seg_modif.nii.gz | 186:210 |   | 55.0 | 74.49899710166940 | 2.0654612628143000 | -0.4655824057061610 | 0.6187649443692570 | 4.602201395154190 | 0.7021559581881710 | 8.21724057257737 | 0.19672289786466700 | 11.508958552747000 | 0.09988903157047450 | 0.699731378857051 | 0.01719526973665410 | 1.219835288456290 | 0.9493392141345070 | 0.9747126663782220 | 0.009245100001059110 | 20.06805395574310


For convenience, here is the original segmentation: 
[t2_seg.nii.gz](https://github.com/spinalcordtoolbox/spinalcordtoolbox/files/6689746/t2_seg.nii.gz)

And the modified segmentation: 
[t2_seg_modif.nii.gz](https://github.com/spinalcordtoolbox/spinalcordtoolbox/files/6689748/t2_seg_modif.nii.gz)

My suggestion would be to leverage the algorithms behind the [`get_centerline()`](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/master/spinalcordtoolbox/centerline/core.py#L86) method, which includes smoothing and extrapolation.

Reply:
thank you for testing this, I'll look into extrapolation of the centerline 🙂

Reply:
I've implemented the extrapolation of the centerline in https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3429/commits/df7cbc986444e7e90feb14868c1b6ef808f244c6

I tested out setting the parameter `param.minmax = None` in `get_centerline()` to directly extrapolate the centerline while generating it. However, with the modified segmentation you did, the result of the extrapolated centerline is really bad. So, what I've comme up with is that I extrapolate on the centerline that is generated only on the slices available in the segmentation minus the top slices of the centerline to exclude those biased values of the centerline.
I just wonder if this would really be generalizable (how many slices, etc.)? It is usually the case that top slices of the segmentation aren't good, so the top of centerline is also biased. Also, even if we don't include the top slices for extrapolation, we still have the general direction of the centerline and that may be enough here
After this, I was a bit out of ideas 😬, I tested out other methods, but the modified segmentation always had a big impact on extrapolation...

PS: The extrapolation was good on the original segmentation